primitives/primproc/batchprimitiveprocessor.cpp

```diff
@@ -541,31 +589,171 @@ void BatchPrimitiveProcessor::addToJoiner(ByteStream& bs)
         idbassert(joinerNum < joinerCount);
         arr = (JoinerElements*) bs.buf();
 
-// 		cout << "reading " << count << " elements from the bs, joinerNum is " << joinerNum << "\n";
-        for (i = 0; i < count; i++)
+        uint32_t &tJoinerSize = tJoinerSizes[joinerNum];
+
+        // XXXPAT: enormous if stmts are evil.  TODO: move each block into
+        // properly-named functions for clarity.
+        if (typelessJoin[joinerNum])
         {
-            if (typelessJoin[joinerNum])
+            vector<pair<TypelessData, uint32_t> > tmpBuckets[processorThreads];
+            TypelessData tlLargeKey;
+            uint8_t nullFlag;
+            PoolAllocator &storedKeyAllocator = storedKeyAllocators[joinerNum];
+            // this first loop hashes incoming values into vectors that parallel the hash tables.
+            for (i = 0; i < count; ++i)
             {
                 bs >> nullFlag;
-
                 if (nullFlag == 0)
                 {
-                    tlLargeKey.deserialize(bs, storedKeyAllocators[joinerNum]);
+                    tlLargeKey.deserialize(bs, storedKeyAllocator);
                     bs >> tlIndex;
-                    tlJoiners[joinerNum]->insert(pair<TypelessData, uint32_t>(tlLargeKey,
-                                                 tlIndex));
+                    bucket = bucketPicker((char *) tlLargeKey.data, tlLargeKey.len, bpSeed) & ptMask;
+                    tmpBuckets[bucket].push_back(make_pair(tlLargeKey, tlIndex));
                 }
                 else
-                    tJoinerSizes[joinerNum]--;
+                    --tJoinerSize;
+            }
+
+            bool done = false, didSomeWork;
+            //uint loopCounter = 0, noWorkCounter = 0;
+            // this loop moves the elements from each vector into its corresponding hash table.
+            while (!done)
+            {
+                //++loopCounter;
+                done = true;
+                didSomeWork = false;
+                for (i = 0; i < processorThreads; ++i)
+                {
+                    if (!tmpBuckets[i].empty())
+                    {
+                        bool gotIt = addToJoinerLocks[joinerNum][i].try_lock();
+                        if (!gotIt)
+                        {
+                            done = false;   // didn't get it, don't block, try the next bucket
+                            continue;
+                        }
+                        for (auto &element : tmpBuckets[i])
+                            tlJoiners[joinerNum][i]->insert(element);
+                        addToJoinerLocks[joinerNum][i].unlock();
+                        tmpBuckets[i].clear();
+                        didSomeWork = true;
+                    }
+                }
+                // if this iteration did no useful work, everything we need is locked; wait briefly
+                // and try again.
+                if (!done && !didSomeWork)
```

Comment:
Not sure what you have in mind.  Do you mean using those instead of a 'back-off' timer when no work could be done?  There are a lot of ways to improve this if we want.  I would suggest tickets with ideas for now, which we can go back to after we've ticked the boxes on the release.